### PR TITLE
Copy the auth user on updateCurrentUser(), also assert the app name matches

### PR DESCRIPTION
**Pull request by samhorlbeck**
This PR creates a differentiation between the public `updateCurrentUser()` and the internal `_updateCurrentUser()`. The latter does not make a copy because it is called by the SDK itself when the SDK signs someone in or updates something.

b/170886559,
b/170854629

### CODE REVIEW

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts [hunk 1]
```diff
@@ -211,7 +211,22 @@ export class AuthImpl implements Auth, _FirebaseService {
...
+    const user = userExtern as User | null;
+    assert(
+      !user || user.auth.name === this.name,
+      AuthErrorCode.ARGUMENT_ERROR,
+      {
+        appName: this.name
+      }
+    );
+
+    return this._updateCurrentUser(user && user._clone());
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/core/auth/auth_impl.ts:226]: user?._clone()

File: packages-exp/auth-exp/src/model/user.ts [hunk 2]
```diff
@@ -34,16 +34,16 @@ export interface UserParameters {
...
   stsTokenManager: StsTokenManager;
 
-  displayName?: string;
-  email?: string;
-  phoneNumber?: string;
-  photoURL?: string;
-  isAnonymous?: boolean;
-  emailVerified?: boolean;
-  tenantId?: string;
+  displayName?: string | null;
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/model/user.ts:37]: we really should figure out this undefined | null situation at some point...

**Review by avolkovi — APPROVED** [review]

File: packages-exp/auth-exp/src/core/auth/auth_impl.ts (hunk 1 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/auth/auth_impl.ts:226]: That types out to `User|undefined` but the method wants `User|null`. So it's either `user?._clone() || null` or `user && user._clone()`. The latter seems cleaner to me so I'm keeping it the way it's written

File: packages-exp/auth-exp/src/model/user.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/model/user.ts:37]: Agreed :P

**Pull request merged**